doc/sphinx/addendum/rewrite-rules.rst

```diff
@@ -0,0 +1,178 @@
+   It is disabled by default and cannot be disabled once enabled.
+   The command-line flag ``-allow-rewrite-rules`` enables rewrite rules at
+   startup.
+
+   .. exn:: Rewrite rule declaration requires enabling the flag "Allow Rewrite Rules".
+      :undocumented:
+
+Symbols
+-----------------
+
+Rewrite rules operate on symbols, which are their own kind of constants.
+They stand in-between defined constants and axioms,
+in that they don't always reduce as defined constants do,
+but they may still reduce using the provided rules, unlike axioms.
+
+.. cmd:: {| Symbol | Symbols } {| {+ ( @assumpt ) } | @assumpt }
+   :name: Symbol; Symbols
+
+   As a command, `Symbol` binds an :n:`ident` to a :n:`type` as a symbol.
+
+   .. coqtop:: in
+
+      Set Allow Rewrite Rules.
+      Symbol pplus : nat -> nat -> nat.
+      Notation "a ++ b" := (pplus a b).
+
+Rewrite rules
+---------------
+
+  .. insertprodn rewrite_rule rewrite_rule
+
+  .. prodn::
+     rewrite_rule ::= {? @univ_decl %|- } @rw_pattern ==> @term
+
+Rewrite rules have two parts simply named left-hand side or pattern and right-hand side.
+When a rule is applied, the term is matched against the pattern,
+subterms aligned with pattern variables are collected
+and substituted into the right-hand side which is returned.
+
+.. cmd:: Rewrite {| Rule | Rules } @ident := {? %| } {+| @rewrite_rule }
+  :name: Rewrite Rule; Rewrite Rules
+
+   The command declares a named block of rewrite rules.
+
+  .. coqtop:: all
+
+     Rewrite Rule pplus_rewrite :=
+     | ?n ++ 0 ==> ?n
+     | ?n ++ S ?m ==> S (?n ++ ?m)
+     | 0 ++ ?n ==> ?n
+     | S ?n ++ ?m ==> S (?n ++ ?m).
+
+Pattern syntax
+--------------
+
+Patterns are a subset of terms which follow a more rigid syntax.
+It is easier to understand them as being a rigid head-pattern, inside of an elimination context:
+
+  .. prodn::
+     rw_head_pattern ::= @qualid {? @univ_annot }
+     | fun ({+ @name } {? : @rw_pattern_arg}) => @rw_pattern_arg
+     | forall ({+ @name } {? : @rw_pattern_arg}), @rw_pattern_arg
+     elim_context ::= []
+     | @elim_context {+ @rw_pattern_arg }
+     | @elim_context .( @qualid {? @univ_annot } )
+     | match @elim_context {? as @name } {? in @pattern } {? return @rw_pattern_arg } with {*| @pattern => @rw_pattern_arg } end
+     rw_pattern ::= @elim_context[@rw_head_pattern]
+     rw_pattern_arg ::= ?@name | _ | @rw_pattern
+
+where :n:`@qualid {? @univ_annot }` (in the first line) can refer to bound variables, symbols, sorts, inductives and constructors, but not arbitrary constants.
+The projections must be primitive to be allowed.
+
+In a few words, patterns are terms with pattern variables (:n:`?@name`),
+but those may not appear on the left of applications or as the discriminee of a match or a primitive projection;
+furthermore a pattern may not have let-bindings or non-symbol constants.
+
+Finally, a valid pattern needs its head head-pattern to be a symbol.
+
+
+Right-hand sides
+----------------
```

Comment:
```suggestion
```

Let's not have a one-sentence section here.  Move the sentence up by the Rewrite command description or drop if already covered.

Reply:
Done